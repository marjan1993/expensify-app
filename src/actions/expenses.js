import uuid from "uuid";

//ADD_EXPENSE
export const addExpense = ({ description = "", note = "", amount = 0, createdAt = 0 } = {}) => ({
    type: "ADD_EXPENSE",
    expense: {
    id: uuid(),
    description,//description : description, but we use shorthand way since the names are the same
    note,
    amount,
    createdAt
    }
});

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({ 
    type: "REMOVE_EXPENSE",
    id  //OR id: id 
});

//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
   type: "EDIT_EXPENSE",
   id,
   updates
});