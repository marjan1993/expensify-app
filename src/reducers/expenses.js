//Expense Reducer 

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case "ADD_EXPENSE":
          //return state.concat(action.expense);
          return [
            ...state, //...state means all of the existing items from the array
            action.expense
          ];
        case "REMOVE_EXPENSE":
          return state.filter(({ id }) => { //shorthand way syntax -->> return state.filter(({ id }) => id !== action.id);
             return id !== action.id 
// if it not equal this will result true, meaning the item will be kept 
// and if it is equal result will be false, meaning the item will be filter out
        });
//(expense) --> we want id so ({ id }) code in filter({ id }) will be better
        case "EDIT_EXPENSE":
          return state.map((expense) => {
             if (expense.id === action.id) {
                return {
                    ...expense,
                    ...action.updates
                };
             } else {
                 return expense;
             };
          });
        default:
           return state;
    };
};


//export default expensesReducer; //old way , we use new way above to export function