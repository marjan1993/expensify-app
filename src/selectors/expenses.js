import moment from "moment";

//get visible expenses

export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
      const createdAtMoment = moment(expense.createdAt)
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true;//new way
      // const startDateMatch = typeof startDate !== "number" || expense.createdAt >= startDate ;// OR || //--> old way
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;//new way
      // const endDateMatch = typeof endDate !== "number" || expense.createdAt <= endDate;//---> old way
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
      //includes() is case-sensitive so we use toLowerCase()
  
     return startDateMatch && endDateMatch && textMatch; //AND &&
     //if all of this are true the filter function will be true and the item will be kept in the array
     // and if any of these are false the whole we be false because we use AND logical so the item will be removed from the array
    }).sort((a, b) => {
       if (sortBy === "date") {
         return a.createdAt < b.createdAt ? 1 : -1 ;//if a < b so b should comes first :(otherwise if not) a will comes first
       } else if (sortBy === "amount") {
           return a.amount < b.amount ? 1 : -1 ;
       } 
     //in code (sortBy === "amount") : return a.amount < b.amount ? -1 : 1 ;if a < b so a should comes first :(otherwise if not) b will comes first
     // with -1 a would come fist
     // with  1 b would come fist
 
    });
 };
 