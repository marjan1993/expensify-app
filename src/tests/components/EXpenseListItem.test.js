import React from "react";
import { shallow } from "enzyme";
import ExpensesListItem from "../../components/ExpenseListItem";
import expenses from "../fixtures/expenses";

test("should render ExpenseListItem correctly", () => {
    const wrapper = shallow(<ExpensesListItem expenses={expenses[0]} />);
    expect(wrapper).toMatchSnapshot();
});