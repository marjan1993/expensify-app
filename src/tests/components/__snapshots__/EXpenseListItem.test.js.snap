// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem correctly 1`] = `
<div>
  <Link
    to="/edit/undefined"
  >
    <h3 />
  </Link>
  <p>
    $0.00
    -
    January 6th, 2020
  </p>
</div>
`;
